Src/ILGPU/Runtime/Cuda/CudaAsm.cs

```diff
@@ -0,0 +1,60 @@
+﻿// ---------------------------------------------------------------------------------------
+//                                        ILGPU
+//                        Copyright (c) 2016-2020 Marcel Koester
+//                                    www.ilgpu.net
+//
+// File: CudaAsm.cs
+//
+// This file is part of ILGPU and is distributed under the University of Illinois Open
+// Source License. See LICENSE.txt for details
+// ---------------------------------------------------------------------------------------
+
+using ILGPU.Frontend.Intrinsic;
+using System;
+using System.Runtime.CompilerServices;
+
+namespace ILGPU.Runtime.Cuda
+{
+    /// <summary>
+    /// Provides library calls for inline PTX assembly instructions.
+    /// </summary>
+    public static partial class CudaAsm
+    {
+        /// <summary>
+        /// Returns true if this running on a Cuda accelerator.
+        /// </summary>
+        public static bool IsSupported
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => Accelerator.CurrentType == AcceleratorType.Cuda;
+        }
+
+        /// <summary>
+        /// Determines if the PTX instruction set is at least the specified version.
+        /// </summary>
+        /// <param name="instructionSet">The PTX instruction set.</param>
+        /// <returns>Return true if the PTX version is supported.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsPTXVersionAtLeast(CudaInstructionSet instructionSet) =>
+            IsPTXVersionAtLeast(instructionSet.Major, instructionSet.Minor);
+
+        /// <summary>
+        /// Determines if the PTX instruction set is at least the specified version.
+        /// </summary>
+        /// <param name="major">The PTX major version.</param>
+        /// <param name="minor">The PTX minor version.</param>
+        /// <returns>Return true if the PTX version is supported.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsPTXVersionAtLeast(int major, int minor) =>
+            IsSupported &&
```

Comment:
I think if we change this `&&` implementation to a bitwise `$`, ILGPU should be able to fold this expression in all cases.

Reply:
Unfortunately, using bitwise `&` does not help.

When `Accelerator.CurrentType` and `Accelerator.LanguageVersion` are specialised, they replace the placeholder IR node, but no folding it performed.

Using `IsSupported` as an example, the IR node will be `CompareValue(AcceleratorTypeValue, PrimitiveValue)`. When `Accelerator.CurrentType` is specialised, the `AcceleratorTypeValue` node is replaced by a `PrimitiveValue` - making the condition `CompareValue(PrimitiveValue, PrimitiveValue)` Later, when `UnreachableCodeElimination` is run, it finds that the condition is `CompareValue`, which cannot be folded - it only knows how to fold `PrimitiveValue`.

Reply:
You are right 😢 The aggressive constant-propagation pass in 57f3232228f2c4cea282570a300ffd92002fa412 should help to aggressively fold all expressions known at compile time. Perhaps right now is the right time to integrate this pass into ILGPU.

Reply:
We originally discussed using the constant-propagation pass in the `O2` optimisation because it was expensive. However, after completing the work for `CudaAsm`, I realised that we need a solution for `O0` too. This is because in `O0` the PTX compiler will see all the PTX instructions, including future instructions that it does not know about.

Reply:
Summary of our discussion from the talk-to-dev session: we came up with the idea of limiting the scope of the (expensive) constant propagation pass to certain basic blocks. This helps to improve performance while being sufficient to optimize the expressions mentioned above.